### PR TITLE
[Sparse] Migrate DiagMatrix class

### PR DESCRIPTION
**Pull request by mufeili**
## Description
@isratnisa If you do not mind, I've also migrated the `__repr__` and `__call__` functions you implemented. You will also need to migrate the elementwise ops for `DiagMatrix`
@czkkkkkk I've not migrated the `transpose` functions you implemented.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by isratnisa — CHANGES_REQUESTED** [review]

File: python/dgl/mock_sparse/diag_matrix.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Values of the diagonal matrix
+
+        Returns
+        -------
+        DiagMatrix
+            Diagonal matrix
```
Comment (jermainewang) [inline python/dgl/mock_sparse/diag_matrix.py:51]: Need example

File: python/dgl/mock_sparse/diag_matrix.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    >>> mat.shape
+    (3, 3)
+    >>> mat.nnz
+    3
+
+    Case2: 3-by-5 matrix with scaler diagonal values
+
+    [[1, 0, 0, 0, 0],
+     [0, 1, 0, 0, 0],
+     [0, 0, 1, 0, 0]]
```
Comment (jermainewang) [inline python/dgl/mock_sparse/diag_matrix.py:201]: I believe identity matrix is always square.

File: python/dgl/mock_sparse/diag_matrix.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+            [1., 1.],
+            [1., 1.]])
+    >>> mat.shape
+    (3, 3)
+    >>> mat.nnz
+    3
+    """
+    len_val = min(shape)
+    if d is None:
+        val_shape = (len_val)
```
Comment (jermainewang) [inline python/dgl/mock_sparse/diag_matrix.py:225]: This is meaningless. Do you mean `val_shape = (len_val,)`?

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: python/dgl/mock_sparse/diag_matrix.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        SparseMatrix
+            The converted sparse matrix object
+
+        Example
+        -------
+
+        >>> import torch
+        >>> val = torch.randn(5)
+        >>> mat = diag(val)
+        >>> sp_mat = mat.as_sparse()
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py:117]: Added `print(sp_mat)`

File: python/dgl/mock_sparse/diag_matrix.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        If not specified, it will be inferred from :attr:`val`, i.e.,
+        (N, N). Otherwise, :attr:`len(val)` must be equal to :attr:`min(shape)`.
+
+    Attributes
+    ----------
+    val : torch.Tensor
+        Diagonal of the matrix.
+    shape : tuple[int, int]
+        Shape of the matrix.
+    """
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py:25]: For `DGLGraph`, users never construct a `DGLGraph` object with the class `DGLGraph`. Instead, they use `dgl.graph` or `dgl.hetero_graph` to construct an object.

Similarly, we have `create_from_coo`, `create_from_csr`, `create_from_csc` for `SparseMatrix` and `diag` for `DiagMatrix`.

File: python/dgl/mock_sparse/diag_matrix.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        Diagonal matrix
+
+    Examples
+    --------
+
+    Case1: 5-by-5 diagonal matrix with scaler values on the diagonal
+
+    >>> import torch
+    >>> val = torch.randn(5)
+    >>> mat = diag(val)
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py:150]: added `print(mat)`

File: python/dgl/mock_sparse/diag_matrix.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    >>> mat.shape
+    (5, 5)
+    >>> mat.nnz
+    5
+
+    Case2: 5-by-10 diagonal matrix with scaler values on the diagonal
+
+    >>> val = torch.randn(5)
+    >>> mat = diag(val, shape=(5, 10))
+    >>> mat.shape
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py:166]: Ditto

File: python/dgl/mock_sparse/diag_matrix.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    3
+
+    Case2: 3-by-5 matrix with scaler diagonal values
+
+    [[1, 0, 0, 0, 0],
+     [0, 1, 0, 0, 0],
+     [0, 0, 1, 0, 0]]
+
+    >>> mat = identity(shape=(3, 5))
+    >>> mat.val
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py:236]: Added `print(mat)`

File: docs/source/api/python/dgl.sparse_v0.rst [hunk 9]
```diff
@@ -13,4 +13,17 @@ Sparse matrix class
 .. currentmodule:: dgl.mock_sparse
 
 .. autoclass:: SparseMatrix
-	:members:
\ No newline at end of file
+	:members:
```
[the remark this answers is not in the document]
Comment (mufeili) [reply docs/source/api/python/dgl.sparse_v0.rst]: done

File: python/dgl/mock_sparse/diag_matrix.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py:51]: updated

File: python/dgl/mock_sparse/diag_matrix.py (hunk 2 above)
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py:230]: Not necessarily, e.g., when adding self-loops for a `DGLBlock`.

File: python/dgl/mock_sparse/diag_matrix.py (hunk 3 above)
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py:257]: They are equivalent for `torch.ones`.

Comment (jermainewang) [reply python/dgl/mock_sparse/diag_matrix.py:257]: But the code doesn't read well. `(variable)` is not creating a tuple. Either you do `val_shape = len_val` or `val_shape = (len_val,)`. Pls change.

File: python/dgl/mock_sparse/diag_matrix.py (hunk 2 above)
Comment (jermainewang) [reply python/dgl/mock_sparse/diag_matrix.py:230]: > e.g., when adding self-loops for a `DGLBlock`.

I think it's ok to have such support but want to point out that maybe the easiest way to add selfloop to `DGLBLock` is using operation like `A[I, I] = val`.

**Review by jermainewang — APPROVED** [review]

File: python/dgl/mock_sparse/diag_matrix.py (hunk 2 above)
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py:230]: Sure.

File: python/dgl/mock_sparse/diag_matrix.py (hunk 3 above)
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py]: Addressed

File: python/dgl/mock_sparse/diag_matrix.py [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+    ----------
+    val : torch.Tensor
+        Diagonal of the matrix.
+    shape : tuple[int, int]
+        Shape of the matrix.
+    """
+    def __init__(self, val: torch.Tensor, shape: Optional[Tuple[int, int]] = None):
+        len_val = len(val)
+        if shape is not None:
+            assert len_val == min(shape), 'Expect len(val) to be min(shape), got '\
```
Comment (frozenbugs) [inline python/dgl/mock_sparse/diag_matrix.py:29]: also use f-string?

File: python/dgl/mock_sparse/diag_matrix.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+    Case3: 5-by-5 diagonal matrix with tensor values on the diagonal
+
+    >>> val = torch.randn(5, 3)
+    >>> mat = diag(val)
+    >>> mat.shape
+    (5, 5)
+    >>> mat.nnz
+    5
+    """
+    # Mufei: this may not be needed if DiagMatrix is simple enough
```
Comment (frozenbugs) [inline python/dgl/mock_sparse/diag_matrix.py:180]: Is this a TODO? or a comment or something else.

Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py:180]: It's a comment.

File: python/dgl/mock_sparse/diag_matrix.py (hunk 10 above)
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py]: Done

File: python/dgl/mock_sparse/diag_matrix.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/dgl/mock_sparse/diag_matrix.py:25]: I see.  Also, noticed that you added examples for `D = DiagMatrix(...)` too. I think now `DiagMatrix` and `SparseMatrix` are consistent with each other in terms of docstring.

File: python/dgl/mock_sparse/diag_matrix.py [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+     [0, 1, 0, 0, 0],
+     [0, 0, 1, 0, 0]]
+
+    >>> mat = identity(shape=(3, 5))
+    >>> print(mat)
+    DiagMatrix(val=tensor([1., 1., 1.]),
+               shape=(3, 5))
+    >>> mat.val
+    tensor([1., 1., 1.])
+    >>> mat.shape
```
Comment (isratnisa) [inline python/dgl/mock_sparse/diag_matrix.py:237]: Are these examples redundant when we `print(mat)` already printing `shape` and `nnz`?

**Review by isratnisa — APPROVED** [review]
Thanks for all the fixes in `sp_matrix.py`. I left some comments for the docstring. Otherwise, LGTM.

File: python/dgl/mock_sparse/diag_matrix.py (hunk 12 above)
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py:237]: Removed

**Comment by mufeili** [comment]
@isratnisa 

> I see.  Also, noticed that you added examples for `D = DiagMatrix(...)`. I think now `DiagMatrix` and `SparseMatrix` are consistent with each other in terms of docstring.

I don't think we need to have examples for `DiagMatrix` and `SparseMatrix`. We also don't have examples for `DGLGraph` as in the doc here: https://docs.dgl.ai/api/python/dgl.DGLGraph.html

**Comment by mufeili** [comment]
@Rhett-Ying Do you have any clues about this failure in C++ CPU?

File: python/dgl/mock_sparse/diag_matrix.py (hunk 11 above)
Comment (jermainewang) [reply python/dgl/mock_sparse/diag_matrix.py:171]: Then you should use `# NOTE(Mufei): xxx.`

**Comment by Rhett-Ying** [comment]
> @Rhett-Ying Do you have any clues about this failure in C++ CPU?

I think it's because I abort the job manually. I found a newer job for this PR is waiting for run, so I stop the previous queued one. And we have some other jobs are queued due to lack of executors. Sorry for confusing you...

**Comment by mufeili** [comment]
> > @Rhett-Ying Do you have any clues about this failure in C++ CPU?
> 
> I think it's because I abort the job manually. I found a newer job for this PR is waiting for run, so I stop the previous queued one. And we have some other jobs are queued due to lack of executors. Sorry for confusing you...

Got it. Thanks.

File: python/dgl/mock_sparse/diag_matrix.py (hunk 11 above)
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py]: Done.

File: python/dgl/mock_sparse/diag_matrix.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/mock_sparse/diag_matrix.py:25]: > Also, noticed that you added examples for D = DiagMatrix(...). I think now DiagMatrix and SparseMatrix are consistent with each other in terms of docstring.

Which ones are you referring to?

**Pull request merged**